### PR TITLE
spread(lxd): work around SSH password authentication disabled by livecd-rootfs

### PR DESCRIPTION
**Pull request by bboozzoo**
In recent images, livecd-rootfs may be disabling SSH password authentication by default, with the intention that cloud-init would enabled that by placing a higher priority file under /etc/sshd/sshd_config.d/. See https://git.launchpad.net/livecd-rootfs/commit/live-build/ubuntu-cpc/hooks.d/chroot/052-ssh_authentication.chroot?id=480d5b26ea97e0bfebac8aedfa1f2bf7286f027a for reference.

Do the same as cloud-init would a drop a higher priority config overriding relevant SSHD settings.

### CODE REVIEW

**Review by zyga — APPROVED** [review]
This looks good to me.

**Comment by bboozzoo** [comment]
I'll extend the lxd backend spread test to cover more recent images.

**Comment by bboozzoo** [comment]
LXD backend spread tests have been updated with Ubuntu 18.04, 20.04, and 22.04.

**Comment by bboozzoo** [comment]
Looks like a similar fix but done slightly differently was proposed in #179

**Comment by thp-canonical** [comment]
Also related, a similar fix for the Google backend: #155

**Review by ZeyadYasser — COMMENTED** [review]
Thank you for catching this, Maybe let's wait for it to be merged in the google backend https://github.com/snapcore/spread/pull/155? and then backport it to all other providers so that it is consistent across all of them.

**Review by cmatsuoka — APPROVED** [review]
Thanks!

**Review by niemeyer — CHANGES_REQUESTED** [review]
Thanks for the PR. Seems reasonable, but a bit obscure. More details below.

File: spread/lxd.go [hunk 1]
```diff
@@ -482,6 +482,8 @@ func (p *lxdProvider) serverJSON(name string) (*lxdServerJSON, error) {
 func (p *lxdProvider) tuneSSH(name string) error {
 	cmds := [][]string{
 		{"sed", "-i", `s/^\s*#\?\s*\(PermitRootLogin\|PasswordAuthentication\)\>.*/\1 yes/`, "/etc/ssh/sshd_config"},
+		// provide a higher priority drop in with our overrides
+		{"/bin/bash", "-c", `[ -d /etc/ssh/sshd_config.d ] && echo -e "PermitRootLogin yes\nPasswordAuthentication yes" > /etc/ssh/sshd_config.d/00-spread.conf`},
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply spread/lxd.go:492]: I have tweaked the wording around the snippets.

[the remark this answers is not in the document]
Comment (cmatsuoka) [reply spread/lxd.go:497]: Thanks for the update. It's better not to force-push after the review started to preserve the dialogue in PR comments (the final landing is squashed so intermediate commits won't appear in the git history anyway). That said, I think the comment wording is still a bit confusing. Let me try an alternative.

**Review by niemeyer — APPROVED** [review]
Thanks!

**Comment by niemeyer** [comment]
Oops, tests are failing.

File: spread/lxd.go [hunk 2]
```diff
@@ -486,7 +486,14 @@ func (p *lxdProvider) serverJSON(name string) (*lxdServerJSON, error) {
...
 func (p *lxdProvider) tuneSSH(name string) error {
 	cmds := [][]string{
+		// Attempt to enable root login with a password through SSH, which is
+		// achieved in a different manner depending on the version of sshd in
+		// the system. Older versions of ssh used a single file.
 		{"sed", "-i", `s/^\s*#\?\s*\(PermitRootLogin\|PasswordAuthentication\)\>.*/\1 yes/`, "/etc/ssh/sshd_config"},
+		// If the sshd configuration is split in snippets in /etc/ssh/sshd_config.d,
+		// place the configuration in a 00-* file because the first obtained value
+		// will be used. See sshd_config(5) for details.
+		{"/bin/bash", "-c", `[ -d /etc/ssh/sshd_config.d ] && echo -e "PermitRootLogin yes\nPasswordAuthentication yes" > /etc/ssh/sshd_config.d/00-spread.conf`},
```
Comment (bboozzoo) [inline spread/lxd.go:496]: The 'shortened' version is actually wrong, exit code `[ -d /etc/ssh/sshd_config.d ]` becomes the return status of the whole command. Let me push a fix.

**Comment by bboozzoo** [comment]
I don't think this is getting anywhere given the flakiness in lxd:
```                                        
Error: Failed instance creation: write unix @->/var/snap/lxd/common/lxd/unix.socket: i/o timeout
-----
```

**Pull request merged**